File: android/app/src/main/java/io/ionic/starter/utils/DateTime.kt
package io.ionic.starter.utils

import android.util.Log
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object DateTime
{

    private val TAG="DateTime"

    fun getTimeZoneOffsetInHours(): Int {
        val timeZone = TimeZone.getDefault()
        val offsetMillis = timeZone.rawOffset + if (timeZone.inDaylightTime(Date())) timeZone.dstSavings else 0
        val offsetHours = offsetMillis / 3600000
        return offsetHours
    }
    val currentDateTimeActual: String
        get()
        {
            val sdf = SimpleDateFormat(
                    "yyyy-MM-dd'T'HH:mm:ss'Z'",
                    Locale.ENGLISH
            )

            sdf.timeZone = TimeZone.getTimeZone("UTC")
            return sdf.format(Calendar.getInstance().time)
        }


    val currentDateTime: String
        get()
        {
            val sdf = SimpleDateFormat(
                    "yyyy-MM-dd'T'HH:mm:ss'Z'",
                    Locale.ENGLISH
            )

            sdf.timeZone = TimeZone.getTimeZone("UTC")

            // Get the current date
            val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
            var days = 0
            val hours =  24-getTimeZoneOffsetInHours()
            if (hours<24)
                days = -1

            calendar.set(Calendar.HOUR_OF_DAY, hours)
            calendar.set(Calendar.MINUTE, 0)
            calendar.set(Calendar.SECOND, 0)
            calendar.set(Calendar.MILLISECOND, 0)
            calendar.add(Calendar.DAY_OF_YEAR,days)

            return sdf.format(calendar.time)
        }





    fun currentAMPM(): String {


        runCatching {


            return SimpleDateFormat("h a",  Locale.getDefault()).format(Date())
        }
            .onFailure {
                Log.i(
                        TAG,
                        "Failed to parse date convertToAmPm ${it.message}"
                )
                it.printStackTrace()
            }

        return "";

    }

    fun convertToAmPm(dateString: String?): String {

        if(dateString==null)
            return ""
        runCatching {

            val sdf = SimpleDateFormat(
                    "yyyy-MM-dd'T'HH:mm:ss'Z'",
                    Locale.ENGLISH
            )
            val date = parseDate(dateString)!!
            val calendar = Calendar.getInstance()
            calendar.time = date
            calendar.add(Calendar.HOUR_OF_DAY, getTimeZoneOffsetInHours())

            return SimpleDateFormat("h a",  Locale.getDefault()).format(calendar.time)
        }
            .onFailure {
                Log.i(
                        TAG,
                        "Failed to parse date convertToAmPm ${it.message}"
                )
                it.printStackTrace()
            }

        return "";

    }

    fun toDayName(dateString: String?): String {

        if(dateString==null)
            return ""
        runCatching {
            val date = parseDate(dateString)!!
            val calendar = Calendar.getInstance()
            calendar.time = date
            calendar.add(Calendar.HOUR_OF_DAY, getTimeZoneOffsetInHours())
            return  SimpleDateFormat("EEE",  Locale.getDefault()).format(calendar.time)
        }
            .onFailure {
                Log.i(
                        TAG,
                        "Failed to parse date convertToAmPm ${it.message}"
                )
                it.printStackTrace()
            }

        return "";

    }
    private fun parseDate(dateString:String): Date?
    {
        kotlin.runCatching {

            return  SimpleDateFormat(
                "yyyy-MM-dd'T'HH:mm:ss'Z'",
                Locale.ENGLISH
        ).parse(dateString)

        }
        return Date()
    }
    fun toShortDate(dateString: String?): String {

        if(dateString==null)
            return ""
        runCatching {

            val date = parseDate(dateString)!!
            val calendar = Calendar.getInstance()
            calendar.time = date
            calendar.add(Calendar.HOUR_OF_DAY, getTimeZoneOffsetInHours())
            val outputFormatter = SimpleDateFormat("dd/MM",  Locale.getDefault())
            return outputFormatter.format(calendar.time)
        }
            .onFailure {
                Log.i(
                        TAG,
                        "Failed to parse date convertToAmPm ${it.message}"
                )
                it.printStackTrace()
            }

        return "";

    }
    fun addDays(days:Int) : String
    {
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.ENGLISH)
        val date=sdf.parse(currentDateTime)?:Date()

        val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        calendar.time=date
        calendar.add(Calendar.DAY_OF_YEAR, days)


        return sdf.format(calendar.time)

    }
}

File: ios/App/App/utils/DateTime.swift
//
//  DateTime.swift
//  App
//
//  Created by Shahzaib Ali on 02/08/2024.
//

import Foundation


class DateTime {

    static let shared = DateTime()
    private let TAG = "DateTime"

    
    func getTimeZoneOffset() -> Int {
        let timeZone = TimeZone.current
        let secondsFromGMT = timeZone.secondsFromGMT()
        let hours = secondsFromGMT / 3600
        return hours
    }
    var currentDateTimeActual: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        formatter.timeZone = TimeZone(abbreviation: "UTC")
        formatter.locale = Locale(identifier: "en")
        return formatter.string(from: Date())
    }

    var currentDateTime: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        formatter.timeZone = TimeZone(abbreviation: "UTC")
        formatter.locale = Locale(identifier: "en")

        // Get the current date
        let currentDate = Date()
        
        // Use Calendar to get the start of the day
        let startOfDay = Calendar.current.startOfDay(for: currentDate)
        
        return formatter.string(from: startOfDay)
    }

    
    private func toAdjustedOffsetHours(dateString: String?) -> Date
    {
        guard let dateString = dateString, let date = parseDate(dateString: dateString) else {
            return  Date()
        }

        // Get the time zone offset in hours
        let offSetHours = getTimeZoneOffset()
        
        // Create a calendar to manipulate the date
        var calendar = Calendar.current
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!

        // Adjust the date with the offset
        let adjustedDate = calendar.date(byAdding: .hour, value: offSetHours, to: date)
        return adjustedDate ?? date
    }
    
    func currentAmPm() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "h a"
        return formatter.string(from:Date())
    }
    func convertToAmPm(dateString: String?) -> String {
     
        let date = toAdjustedOffsetHours(dateString: dateString)
        let formatter = DateFormatter()
        formatter.dateFormat = "h a"
//        formatter.locale = Locale(identifier: "en_US") // You can use a different locale if needed
        return formatter.string(from:date)
    }

    func toDayName(dateString: String?) -> String {
        let date = toAdjustedOffsetHours(dateString: dateString)
        let formatter = DateFormatter()
        formatter.dateFormat = "EEE"
//        formatter.locale = Locale(identifier: "time_zone".toLocalizedString())
        return formatter.string(from: date)
    }

    func toShortDate(dateString: String?) -> String {
        let date = toAdjustedOffsetHours(dateString: dateString)
        let formatter = DateFormatter()
        formatter.dateFormat = "dd/MM"
//        formatter.locale = Locale(identifier: "time_zone".toLocalizedString())
        return formatter.string(from: date)
    }

    func addDays(days: Int) -> String {
        let formatter = DateFormatter()
           formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
           formatter.timeZone = TimeZone(abbreviation: "UTC")
           formatter.locale = Locale(identifier: "en")
           let calendar = Calendar.current
           
           // Get the current date
           let currentDate = Date()
           
           // Get the start of the current day
           let startOfDay = calendar.startOfDay(for: currentDate)
           
           // Add the specified number of days to the start of the current day
           let newDate = calendar.date(byAdding: .day, value: days, to: startOfDay)!
           
           return formatter.string(from: newDate)
    }

    private func parseDate(dateString: String) -> Date? {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
//        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter.date(from: dateString)
    }
}
